File: Common/SocketMessages.py
import logging
from abc import abstractmethod
from typing import List
from flask_socketio import emit

logger = logging.getLogger(__name__)

class Message:
    _errorKey = "error"
    _dataKey =  "data"

    def __init__(self, message:str):
        self._messageContent={}
        self.data=None
        self.message = message

    @abstractmethod
    def _setMessage(self, data):
        pass

    def sendMessage(self, data):
        self._setMessage(data)
        self._addMessageToContent()
        self._emitMessage()

    def sendError(self, error:str):
        self._messageContent[self._errorKey] = error
        self._emitMessage()

    def _addMessageToContent(self):
        self._messageContent[self._dataKey] = self.data

    def _emitMessage(self):
        emit(self.message, self._messageContent)


# ------------------ MediaInfo_response -----------------------------------
class MediaInfo:
    def __init__(self, title:str=None, artist:str=None):
        self.title = title
        self.artist = artist

# data: {"artist", "title"}
class MediaInfo_response(Message):
    message = "getMediaInfo_response"

    def __init__(self):
        super().__init__(self.message)

    def _setMessage(self, mediaInfo:MediaInfo):
        self.data = {}
        if mediaInfo.artist is not None:
            self.data["artist"] = mediaInfo.artist
        else:
            self.data["artist"] = " "
        self.data["title"] = mediaInfo.title


# ------------------- downloadMedia_finish -------------------------------
# data: hash
class DownloadMedia_finish(Message):
    message="downloadMedia_finish"

    def __init__(self):
        super().__init__(self.message)

    def _setMessage(self, hash:str):
        self.data = hash


# -------------------- getPlaylistInfo_response ----------------------------
class MediaFromPlaylist:
    def __init__(self, index:str, url:str, title:str):
        self.playlistIndex = index
        self.url = url
        self.title = title

class PlaylistInfo:
    def __init__(self, name:str, listOfMedia:List[MediaFromPlaylist]):
        self.playlistName = name
        self.listOfMedia = listOfMedia

# data: playlistName, [{"playlist_index" "url" "title"}, ..]
class PlaylistInfo_response(Message):
    message="getPlaylistInfo_response"

    def __init__(self):
        super().__init__(self.message)

    def _setMessage(self, playlistInfo:PlaylistInfo):
        self.data = []
        listOfMedia = playlistInfo.listOfMedia
        playlist = []
        for media in listOfMedia:
            mediaInfo = {}
            mediaInfo["playlist_index"] = media.playlistIndex
            mediaInfo["url"] = media.url
            mediaInfo["title"] = media.title
            playlist.append(mediaInfo)
        self.data.append(playlistInfo.playlistName)
        self.data.append(playlist)


# ------------------ getPlaylistMediaInfo_response -------------------------
class PlaylistMediaInfo:
    def __init__(self, playlistIndex:str, filename:str, hash:str):
        self.playlistIndex = playlistIndex
        self.filename = filename
        self.hash = hash

# data: {playlist_index, filename, hash}
class PlaylistMediaInfo_response(Message):
    message = "getPlaylistMediaInfo_response"

    def __init__(self):
        super().__init__(self.message)

    def _setMessage(self, playlistInfo:PlaylistMediaInfo):
        self.data = {}
        self.data["playlist_index"] = playlistInfo.playlistIndex
        self.data["filename"]       = playlistInfo.filename
        self.data["hash"]           = playlistInfo.hash


# ------------------- downloadMediaFromPlaylist_start ---------------------------
# data: {playlist_index, filename}
class DownloadMediaFromPlaylist_start(Message):
    message = "downloadMediaFromPlaylist_start"

    def __init__(self):
        super().__init__(self.message)

    def _setMessage(self, playlistInfo:PlaylistMediaInfo):
        self.data = {}
        self.data["playlist_index"] = playlistInfo.playlistIndex
        self.data["filename"]       = playlistInfo.filename


# ------------------- downloadMediaFromPlaylist_finish ---------------------------
# data: {playlist_index, filename}
class DownloadMediaFromPlaylist_finish(Message):
    message = "downloadMediaFromPlaylist_finish"

    def __init__(self):
        super().__init__(self.message)

    def _setMessage(self, playlistInfo:PlaylistMediaInfo):
        self.data = {}
        self.data["playlist_index"] = playlistInfo.playlistIndex
        self.data["filename"]       = playlistInfo.filename


# ------------------- downloadPlaylist_finish -----------------------------
# data: numberOfDownloadedPlaylists
class DownloadPlaylists_finish(Message):
    message = "downloadPlaylist_finish"

    def __init__(self):
        super().__init__(self.message)

    def _setMessage(self, numberOfDownloadedPlaylists:int):
        self.data = numberOfDownloadedPlaylists

File: youtubedlWeb/static/SocketMessages.js

class SocketManager
{
    constructor(socket)
    {
        this.socket = socket;
        socket.on('connect', function() {});
    }

    sendMessage(msg)
    {
        if (!(msg instanceof Message))
        {
            console.error("wrong data for send message")
            return
        }
        this.socket.emit(msg.messageKey, msg._messageContent)
    }
}

class MessageBase
{
    constructor(msg)
    {
        this.message = msg
        this.errorStr = "error"
        this.dataStr = "data"
    }

    isSuccess()
    {
        if (this.dataStr in this.message)
        {
            return true;
        }
        return false;
    }

    isError()
    {
        if(this.errorStr in this.message)
        {
            return true;
        }
        return false;
    }

    _getData(data)
    {
        return this.message[this.dataStr]
    }

    getData(){
        if (this.dataStr in this.message){
            this.data = this.message[this.dataStr]
            return this._getData();
        }
    }

    _getError()
    {
        return this.message[this.errorStr]
    }

    getError()
    {
        if (this.errorStr in this.message)
        {
            this.error = this.message[this.errorStr]
            return this._getError()
        }
    }
}


// ----------------- MediaInfo_response --------------------
class MediaInfo
{
    constructor(title, artist, hash){
        this.title = title;
        this.artist = artist;
        this.hash = hash;
    }
}

// data: {"artist", "title", "hash"}
class MediaInfo_response extends MessageBase{
    static Message = "getMediaInfo_response"
    constructor(msg)
    {
        super(msg)
    }

    _getData()
    {
        this.mediaInfo = new MediaInfo(this.data["title"], this.data["artist"], this.data["hash"])
        return this.mediaInfo
    }
}


// -------------- downloadMedia_finish ------------------------
// data: hash
class DownloadMedia_finish extends MessageBase{
    static Message = "downloadMedia_finish"
    constructor(msg)
    {
        super(msg)
    }
}


// --------------- getPlaylistInfo_response -----------------
class MediaFromPlaylist
{
    constructor(playlistIndex, url, title)
    {
        this.playlistIndex = playlistIndex
        this.url = url
        this.title = title
    }
}

class PlaylistInfo
{
    constructor(playlistName, listOfMedia)
    {
        this.playlistName = playlistName
        this.listofMedia = listOfMedia
    }
}

// data: playlistName, [{"playlist_index" "url" "title"}, ..]
class PlaylistInfo_response extends MessageBase{
    static Message = "getPlaylistInfo_response"
    constructor(msg)
    {
        super(msg)
    }

    _getData()
    {
        var playlistName = this.data[0]
        var playlist = this.data[1]
        var list = []
        for (const element of playlist) {
            list.push(new MediaFromPlaylist(element["playlist_index"], element["url"], element["title"]))
        }
        var playlistInfo = new PlaylistInfo(playlistName, list)
        return playlistInfo
    }
}


// ------------  getPlaylistMediaInfo_response ----------------
class PlaylistMediaInfo{
    constructor(playlistIndex, filename, hash)
    {
        this.playlistIndex = playlistIndex
        this.filename = filename
        this.hash = hash
    }
}

// data: {"playlist_index" "filename" "hash"}
class PlaylistMediaInfo_response extends MessageBase{
    static Message = "getPlaylistMediaInfo_response"
    constructor(msg)
    {
        super(msg)
    }

    _getData()
    {
        return new PlaylistMediaInfo(this.data["playlist_index"], this.data["filename"], this.data["hash"])
    }
}


// --------------  downloadMediaFromPlaylist_start ------------------
// data: {playlist_index, filename}
class DownloadMediaFromPlaylist_start extends MessageBase{
    static Message = "downloadMediaFromPlaylist_start"
    constructor(msg)
    {
        super(msg)
    }

    _getData()
    {
        return new PlaylistMediaInfo(this.data["playlist_index"], this.data["filename"])
    }
}


// --------------  downloadMediaFromPlaylist_finish ------------------
class DownloadMediaFromPlaylistError
{
    constructor(index, ytHash, title)
    {
        this.index = index
        this.ytHash = ytHash
        this.title = title
    }
}

// data: {playlist_index, filename}
class DownloadMediaFromPlaylist_finish extends MessageBase {
    static Message = "downloadMediaFromPlaylist_finish"
    constructor(msg) {
        super(msg)
    }

    _getData() {
        return new PlaylistMediaInfo(this.data["playlist_index"], this.data["filename"])
    }

    _getError() {
        return new DownloadMediaFromPlaylistError(this.error["index"], this.error["ytHash"], this.error["title"])
    }
}


// ----------------  downloadPlaylist_finish ------------------
// data: numberOfDownloadedPlaylists
class DownloadPlaylists_finish extends MessageBase{
    static Message = "downloadPlaylist_finish"
    constructor(msg)
    {
        super(msg)
    }
}
